File: MCTabBarController/MCTabBarController/MCTabBarController.m
//
//  MCTabBarController.m
//  MCTabBarController
//
//  Created by caohouhong on 2018/12/7.
//  Copyright © 2018年 caohouhong. All rights reserved.
//  github:https://github.com/Ccalary/MCTabBarDemo

#import "MCTabBarController.h"

@interface MCTabBarController ()<UITabBarControllerDelegate>
@end

@implementation MCTabBarController

- (void)viewDidLoad {
    [super viewDidLoad];
    _mcTabbar = [[MCTabBar alloc] init];
    [_mcTabbar.centerBtn addTarget:self action:@selector(buttonAction:) forControlEvents:UIControlEventTouchUpInside];
    //利用KVC 将自己的tabbar赋给系统tabBar
    [self setValue:_mcTabbar forKeyPath:@"tabBar"];
    self.delegate = self;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

// 重写选中事件， 处理中间按钮选中问题
- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController{
    
    _mcTabbar.centerBtn.selected = (tabBarController.selectedIndex == self.viewControllers.count/2);
    
    if (self.mcDelegate){
        [self.mcDelegate mcTabBarController:tabBarController didSelectViewController:viewController];
    }
}

- (void)buttonAction:(UIButton *)button{
    NSInteger count = self.viewControllers.count;
    self.selectedIndex = count/2;//关联中间按钮
    if (self.delegate){
        [self.delegate tabBarController:self didSelectViewController:self.viewControllers[self.selectedIndex]];
    }
}
@end

File: MCTabBarControllerSwift/MCTabBarSwift/MCTabBarController.swift
//
//  MCTarBarController.swift
//  MCTabBarControllerSwift
//
//  Created by caohouhong on 2018/12/11.
//  Copyright © 2018年 chh. All rights reserved.
//  https://github.com/Ccalary/MCTabBarController

import UIKit

protocol MCTarBarControllerDelegate : NSObjectProtocol {
     func mcTabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController)
}

class MCTabBarController: UITabBarController,UITabBarControllerDelegate {

    var mcTabBar = MCTabBar(frame: CGRect.zero)
    weak var mcDelegate : MCTarBarControllerDelegate?
    override func viewDidLoad() {
        super.viewDidLoad()
        mcTabBar.centerBtn.addTarget(self, action: #selector(centerBtnAction), for: .touchUpInside)
        self.setValue(mcTabBar, forKeyPath: "tabBar")
        self.delegate = self
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
        mcTabBar.centerBtn.isSelected = (tabBarController.selectedIndex == (viewControllers?.count)!/2)
        self.mcDelegate?.mcTabBarController(tabBarController, didSelect: viewController)
    }
    // 中间按钮点击
    @objc func centerBtnAction() {
        let count = viewControllers?.count ?? 0
        self.selectedIndex = count/2 // 关联中间按钮
        self.tabBarController(self, didSelect: viewControllers![selectedIndex])
    }
}
